### PR TITLE
Bump compileSdkVersion and finalize setAllowActivityStarts API.

### PR DESCRIPTION
**Pull request by jpd236**
Per https://developer.android.com/about/versions/14#beta-3, Android U has reached the platform stability milestone which means that all external APIs are finalized.

We can thus bump the compileSdkVersion to 34 (U) and begin using APIs added there. We leave targetSdkVersion unchanged for now to avoid the broader evaluation of whether deeper changes may be necessary as part of the upgrade; this simply allows compile-time access to newer APIs without changing runtime behavior.

See b/274061424

### CODE REVIEW

**Review by markb74 — APPROVED** [review]

**Review by ejona86 — APPROVED** [review]

**Comment by ejona86** [comment]
Uhh....
> ERROR:/opt/android-sdk/current/platforms/android-34/optional/android.car.jar: R8: java.lang.IllegalArgumentException: Unsupported class file major version 61

Does the new Android require Java 17?

Also... would this require downstream users to also use SDK 34? It seems a bit strange to require them to use the beta SDK.

**Comment by ejona86** [comment]
@markb74, sorry, I didn't mean to dismiss your review. I clicked the button for me, but github reordered us as I was clicking...

**Comment by jpd236** [comment]
Following up internally regarding the compilation error. This feels likely to me to be an oversight as none of the public-facing documentation I can see indicates a change in app-facing requirements, and since it only appears to impact this particular JAR and not the others in the SDK.

I don't _believe_ that this would in turn require downstream clients to use SDK 34, because there's no observable exposure of the compile SDK version in the generated .aar or other release artifacts. There is a file META-INF/com/android/build/gradle/aar-metadata.properties with a minCompileSdk version, but I don't think this is set directly to the compileSdkVersion, as it is showing up as "1" in the most recent version of grpc-binder, not 33.

Even if it did, I don't think it's a problem. General guidance is that upgrading compileSdkVersion is always safe as it does not change runtime behavior (see e.g. https://medium.com/androiddevelopers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd). 

All that said, we should definitely figure out whether Java 17 is actually intended to be a requirement; if it is, then it probably makes sense to defer until there's more critical mass behind it. If it's not, then IMO it's safe, but this isn't urgent given that the copied constant didn't change during SDK finalization, so we could hold off until that "beta" designation is removed officially.

**Comment by jpd236** [comment]
Hi - would it be possible to trigger a retry of the failing (and/or) all checks here? I believe that there is a new revision of the new Android SDK available which should have resolved this issue.

**Review by ejona86 — APPROVED** [review]
Android CI is passing with the new SDK build. Yay!

**Comment by ejona86** [comment]
FYI @temawi

**Pull request merged**
